Add type definitions

TaylorBeeston opened the pull request:
`Closes GH-28`

This PR adds the `types` directory that contains an `index.d.ts` file that allows TypeScript projects to use this package more easily. Additionally, there are some tests to ensure the correctness of this type definition, and some boilerplate associated with dtslint. `package.json` has been updated to include `dtslint` and `unified` (this is required for the `Plugin` type) as explicit dependencies. There is also the addition of the `test-types` script to run dslint, and `index.d.ts` is exported as the types file for this project.

This is my first time writing a `.d.ts` file for a project, so please forgive me if there are any issues. One thing I would like to note is that the `sanitize` parameter is currently just accepting nearly any object. I feel that a more rigid type should be defined within [hast-util-sanitize](https://github.com/syntax-tree/hast-util-sanitize) based upon the definition of `schema` in the readme. If needed, I can try and write a `.d.ts` file for that repo as well, however, I would like to try and solve one step at a time, and make sure that I am doing things correctly along the way.

TaylorBeeston committed:
Explicitly depend on unified for types

TaylorBeeston committed:
Add types

ChristianMurphy reviewed (approved):
Thanks @TaylorBeeston!

ChristianMurphy commented on types/index.d.ts:
```diff
@@ -0,0 +1,11 @@
+import {HastUtilToHtmlOptions} from 'hast-util-to-html'
+
+interface htmlOptions extends HastUtilToHtmlOptions {
+  sanitize?: boolean | {[key: string]: any}
```
This is a good starting point, as mentioned in the description, it is a bit broad.
The preference in Unified/Remark is to get upstream typings in place, then leverage them where it makes sense.

>  If needed, I can try and write a .d.ts file for that repo as well

Thanks! This would be appreciated

TaylorBeeston replied:
No problem! I will go ahead and get started!

ChristianMurphy replied:
Also #33 reminded me there is a `handlers` property, which is based off https://github.com/syntax-tree/mdast-util-to-hast/blob/81ae2dde0ac255d413df0ca7bf20e2ceae1b870e/types/index.d.ts#L12-L22 which we may want to add

TaylorBeeston replied:
Good catch! I had to add type definitions for [hastscript](https://github.com/syntax-tree/hastscript) before I can do the type definitions for hast-util-sanitize, so once those two PRs go through I'll update this one to include both `sanitize` and `handlers`

BarryThePenguin reviewed (approved):
📈 looks good! 👍🏻

TaylorBeeston committed:
Update mdast-util-sanitize to ^3.0.0

TaylorBeeston committed:
Merge branch 'main' into add-type-definitions

TaylorBeeston committed:
Update mdast-util-to-hast to ^9.0.0

TaylorBeeston committed:
Update type definition to include sanitize and handlers

TaylorBeeston commented:
I went ahead and added the correct type definitions for `sanitize` and `handlers`. I also added some tests to reflect the change. 

This is the first time I've had a pull request open while the main branch merged in changes, so I'm not 100% certain I handled it correctly. I just `git pull`ed `main`, then merged my local version of `main` into this branch (see commit 03206a3, "Merge branch 'main' into add-type-definitions"). I am hoping that was the right thing to do!

ChristianMurphy commented on package.json:
```diff
@@ -28,18 +28,21 @@
   "devDependencies": {
     "browserify": "^16.0.0",
     "commonmark.json": "^0.29.0",
+    "dtslint": "^3.6.12",
```
```suggestion
    "dtslint": "^3.0.0",
```

ChristianMurphy reviewed (approved):
LGTM, thanks @TaylorBeeston! :bow:

TaylorBeeston committed:
Use ^3.0.0 for dtslint

Co-authored-by: Christian Murphy <christian.murphy.42@gmail.com>

wooorm commented:
Thanks @TaylorBeeston, released!
